lspAst.sml

```diff
@@ -1,42 +1,113 @@
-    
-    
-    datatype result = Result;
-    datatype errorCode = ParseError | InvalidRequest | MethodNotFound | InvalidParams
-                    | InternalError | ServerErrorStart | ServerErrorEnd
-                    | ServerNotInitialized | UnknownErrorCode | RequestCancelled;
-    type data = string;
-    datatype responseError = ResponseError of errorCode * message * data;
-    datatype error = Error of responseError;
-    datatype ResponseMessage = ResponseMessage of id * result * error;
-    
-    datatype cancelParams = CancelParams of id;
-    type scheme = string;
-    type authority = string;
-    type path = string;
-    type query = string;
-    type fragment = string;
-    datatype uri = DocumentUri of scheme * authority * path * query * fragment;
+	type Id = string;  (* Unique Id identifying a request/response *)
+	type Message = string;
+	datatype Method = CancelRequest | string;  (* Method called for the request *)
+	datatype Params = CancelParams | string;  (* parameters for the method *)
+    type RequestMessage = Id * Method * Params;
+    
+    datatype ErrorCode = ParseError | InvalidRequest | MethodNotFound | InvalidParams
+                   | InternalError | ServerErrorStart | ServerErrorEnd | ServerNotInitialized
+                   | UnknownErrorCode | RequestCancelled;
+    type Data = string  (* structured data about error code *)
+    type ResponseError = ErrorCode * Message * Data;
+    type Result = string; (* result of the response *)
+    type ResponseMessage = Id * Result * ResponseError;
+    
+    type NotificationMessage = Method * Params;
+    
+    type CancelParams = Id;
+    
+    (* uri *)
+    (* parsed into scheme, authority, path, query, and fragment *)
+    type Scheme = string;
+    type Authority = string;
+    type Path = string;
+    type Query = string;
+    type Fragment = string;
+    
+    type DocumentUri = Scheme * Authority * Path * Query * Fragment;
     
     datatype EOL = NEWLINE | CARRIAGE | CARRIAGENEWLINE;
-    type line = int;
-    type character = int;
-    datatype position = POSITION of line * character;
-    datatype range = Range of position * position;
     
-    datatype location = Location of uri * range;
+    type Line = int;
+    type CharPos = int;
+    type Position = Line * CharPos;
+    
+    type Range = Position * Position;
+    
+    type Location = DocumentUri * Range;
+    
+    (* Diagnostic related types *)
+    
+    datatype DiagnosticSeverity = Error | Warning | Information | Hint;
+    datatype DiagnosticCode = string;  (* Diagnostic Code for appearing in UI *)
+    type DiagnosticSource = string;   (* source of the diagnostic - human readable string *)
+    type DiagnosticMessage = string;
+    type DiagnosticRelatedInfoMessage = string;
+    type DiagnosticRelatedInfo = Location * DiagnosticRelatedInfoMessage;
+    type Diagnostic = Range * DiagnosticSeverity * DiagnosticCode * DiagnosticSource * DiagnosticMessage * DiagnosticRelatedInfo;
```

Comment:
Convert this into a record for easier selection of component. Even Range should become a record with fields start and end.